File: ailist/src/main/java/me/kosik/interwalled/ailist/model/Interval.java
package me.kosik.interwalled.ailist.model;

import java.io.Serializable;
import java.util.Objects;

public record Interval<T>(
        String key,
        long from,
        long to,
        T value
) implements Serializable {

    public static boolean overlaps(final Interval<?> lhs, final Interval<?> rhs) {
        return Objects.equals(lhs.key(), rhs.key()) && lhs.to() >= rhs.from() && rhs.to() >= lhs.from();
    }
}

File: ailist/src/main/scala/me/kosik/interwalled/ailist/model/Interval.scala
package me.kosik.interwalled.ailist.model

case class Interval(
  key:    String,
  from:   Long,
  to:     Long,
  value:  String
)

object Interval {
  def overlaps(lhs: Interval, rhs: Interval): Boolean =
    lhs.key ==  rhs.key && lhs.to >= rhs.from && rhs.to >= lhs.from
}
